lib/ecto_adapters_dynamodb/query.ex

```diff
@@ -90,58 +100,89 @@ defmodule Ecto.Adapters.DynamoDB.Query do
         query = construct_search(index, search, opts)
 
         if op == :in do
-          responses_element = "Responses"
-          response_map = %{responses_element => %{table => []}}
+          response_map = %{@responses_element_key => %{table => []}}
 
           Enum.reduce(hash_values, response_map, fn(hash_value, acc) ->
             # When receiving a list of values to query on, construct a custom query for each of those values to pass into do_fetch_recursive/1.
-            %{"Items" => items} = Kernel.put_in(query, [:expression_attribute_values, :hash_key], hash_value)
+            %{"Items" => items} = put_in(query, [:expression_attribute_values, :hash_key], hash_value)
                                   |> (do_fetch_recursive).()
 
-            Kernel.put_in(acc, [responses_element, table], acc[responses_element][table] ++ items)
+            put_in(acc, [@responses_element_key, table], acc[@responses_element_key][table] ++ items)
           end)
         else
           do_fetch_recursive.(query)
         end
       :scan ->
         maybe_scan(table, search, opts)
     end
+  end
 
-    results
+  @spec filter(dynamo_response,  non_indexed_filters, String.t) :: dynamo_response | %{}
+  defp filter(%{"Item" => item} = result, non_indexed_filters, _) do
+    if passes_filter?(item, non_indexed_filters),
+      do: result,
+      else: %{}
   end
+  defp filter(%{@responses_element_key => items} = result, non_indexed_filters, table) do
+    filtered_results = Enum.filter(items[table], &passes_filter?(&1, non_indexed_filters))
+    put_in(result, [@responses_element_key, table], filtered_results)
+  end
+
 
-  # If a primary key query has additional search clauses that are not reflected by the indexes,
-  # we may need to use a Dynamo query instead of get_item in order to apply filters.
-  defp should_query?(_indexes, []), do: false
-  defp should_query?(indexes, [{logical_op, search_clauses} | additional_search_clauses]) when logical_op in @logical_ops,
-    do: should_query?(indexes, search_clauses) or should_query?(indexes, additional_search_clauses)
-  defp should_query?(indexes, [{field, _} | search_clauses]) do
-    if field not in indexes,
-      do: true,
-      else: should_query?(indexes, search_clauses)
+  # @spec passes_filter?(dynamo_response, non_indexed_filters) :: boolean()
+  defp passes_filter?(item, [{ logical_op, [ deeper_clauses, filter ] }]) do
+    case logical_op do
+      :and -> evaluate_filter_expression(item, filter) and passes_filter?(item, deeper_clauses)
+      :or -> evaluate_filter_expression(item, filter) or passes_filter?(item, deeper_clauses)
+    end
   end
+  defp passes_filter?(item, [{ _, [ filter ] }]) do
+    evaluate_filter_expression(item, filter)
+  end
+  defp passes_filter?(item, deeper_clauses) do
```

Comment:
Not sure I get this one. I might not be up to speed on the structures. Why are we matching anything (not matched so far) and nesting it in a list?

Reply:
The third condition is there to handle two possible scenarios where the second argument to the function is a two-element tuple, rather than a keyword list. By nesting that argument in a list, it gets passed over to the to other definitions of `passes_filter?`, depending on other aspects of its structure.

For example, the query

```
lower_bound = 17
upper_bound = 19

from(p in Person,
              where: p.id == "person:jamesholden"
                and p.email == "jholden@expanse.com"
                and p.last_name == "Holden"
                and fragment("? between ? and ?", p.age, ^lower_bound, ^upper_bound),
              select: p.id)
             |> TestRepo.one() == person.id
```

will end up calling that third definition of `passes_filter?` two times - the first time, the second argument is

```
{:and,
 [
   {:and, [{"email", {"jholden@expanse.com", :==}}]},
   {"last_name", {"Holden", :==}}
 ]}
```

where it has both a filter and deeper clauses, so by wrapping that in a list and recursing, it gets passed over to the first definition and processed accordingly.

The second time, it just receives

```
{:and, [{"email", {"jholden@expanse.com", :==}}]}
```

where there's just a filter, so by wrapping it here, it gets passed to the second definition where the filter gets passed off for evaluation.

I'll add additional test cases to make sure this third definition is getting called, and I'll rename the second argument to be more clear that it could be either a set of deeper clauses or a filter.